Increased amount of day/night cycle gradients.

soupjake opened the pull request:
I've added upon the app's functionality of setting the background gradient based on day/night cycles by implementing more stages. The cycle will now go in the order of: midnight, night, twilight, dawn/dusk, morning/eve, day, midday, and then vice versa. These values are based upon the percentage of the current time from the `lastMidnight`, `sunrise`, `sunset` and `nextMidnight` values. 

Here are the background gradients I have chosen:
![](https://i.imgur.com/4jmy5Zk.jpg)

I used just the standard Flutter `Color` class for generating these but if you feel like you want to provide better, more custom colours for these stages then go ahead!

Let me know what you think.

Commit:
Update application_colors.dart

Added more colours for different parts of day/night cycle.

Commit:
Update weather_main_screen.dart

Updated to reflect new ApplicationColors names.

Commit:
Update widget_helper.dart

Updated `buildGradientBasedOnDayCycle` to now set background gradient to more day/night cycles.

Commit:
Update widget_helper.dart

Forgot to change the `getGradient` colours oops.

Commit:
Update application_colors.dart

Fixed typo on `dawnDuskEndColod`

Commit:
Update widget_helper.dart

Changed from twilight to dawn/dusk colours when percentage <= 0.2 || >= 0.8 during day

Commit:
Update widget_helper.dart

Added more steps to include dawn/dusk both before sunrise and after sunset and morning/eve colours.

jhomlala commented:
Hi! Your feature is pretty cool. I will be happy to merge your changes, but I see 2 problems in your commit log. First is "If/else wall" which means that your buildGradientBasedOnDayCycle method contains too much statements. I propose to split it into multiple methods for better code readability. Second, please add tests, because I must be sure that all if cases works. To test code with DateTime.now, I propose adding "currentTime" parameter in method which gives you chance to test your logic.

Anyway good job! Please keep going and I will merge your changes :) 👍

thewithz commented:
Sorry to piggyback off this, but generally when I see "if else chain" it can
almost always be optimized with a switch statement. If you convert the
percent to an int, it should work fine.

EDIT: deleted duplicate content Github adds when replying be email. 
EDIT 2: typos

soupjake commented:
A switch can't be used because all cases being checked are bools and non-const variables, such as the `DateTime` and `sunrise`, `sunset` parameter variables, are being used. I'll try my best at improving it though.

Commit:
Update widget_helper.dart

Improved `buildGradientBasedOnDayCycle` "if/else wall" but using `getNightGradient` and `getDayGradient` methods.

soupjake commented:
Okay improved the "if/else wall", let me know what you think.

Will work on adding tests next.

jhomlala commented:
Please change WidgetHelper.buildGradient() to buildGradient() (remove "WidgetHelper."). Please add tests and I will merge your PR :)

Commit:
Update widget_helper.dart

Removed unnecessary 'WidgetBuilder.' on 'buildGradient' methods.

Commit:
Create gradient_cycle_test.dart

Added test for checking if day/night cycles get the correct gradients based on hour percentages.

soupjake commented:
Okay removed unnecessary "WidgetHelper."s from method and added a test which goes through all the various times of day to see if their respective gradients are returned. Let me know if this is sufficient.

jhomlala commented:
Tests are OK for me, just one thing. Instead writing one test and multiple expect, please create multiple test with one expect, like this:

```
  test("testing mindnight gradient", (){
    expect(testGradientCycle(0, 0), buildGradient(ApplicationColors.midnightStartColor, ApplicationColors.midnightEndColor));
  });
  test("testing night cycle", (){
    expect(testGradientCycle(4, 0), buildGradient(ApplicationColors.nightStartColor, ApplicationColors.nightEndColor));
  });
```
The reason why it's necessary, is because when test fails, It will display test name and we immediately know which test fails.

Anyway everything is OK, you're almost there, just one small fix :)

Commit:
Update gradient_cycle_test.dart

Separated tests.

Commit:
Update gradient_cycle_test.dart

Forgot a word in a comment.

soupjake commented:
Nah yeah my bad for not doing that the first time. Hope that's better now.

jhomlala commented:
Merged :) Thank you for you contribution. Please see issues if you want to add something more :)

soupjake commented:
No problem! Really like the look of this app and thought I'd contribute my idea. Thanks for your patience with my messy programming approach!
